pkg/promotion/releasepromotion/release.go

```diff
@@ -0,0 +1,205 @@
+package releasepromotion
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	cioperatorapi "github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Verifier verifies that a set of ci-operator job definitions have unique image promotions
+// and allows generation of standard files that can be approved for changes for the images
+// that are promoted. It also can generate mirroring jobs.
+type Verifier struct {
+	targets verifierMap
+	errs    []error
+}
+
+// NewVerifer constructs a promotion and mirroring verifier.
+func NewVerifier() *Verifier {
+	return &Verifier{
+		targets: make(verifierMap),
+	}
+}
+
+// Errors returns all errors the verifier has detected.
+func (v *Verifier) Errors() []error {
+	return v.errs
+}
+
+// LoadConfig processes a single ci-operator configuration and returns an error if the config could not
+// be processed. Verification errors are added for later retrieval by Errors().
+func (v *Verifier) LoadConfig(config *cioperatorapi.ReleaseBuildConfiguration, info *config.Info) error {
+	p := config.PromotionConfiguration
+	if p == nil || p.Disabled {
+		return nil
+	}
+	excluded, included := sets.NewString(p.ExcludedImages...), sets.NewString()
+	key := promotionTarget{Namespace: p.Namespace, Name: p.Name, NamePrefix: p.NamePrefix, Tag: p.Tag}
+	if key.Empty() {
+		v.errs = append(v.errs, fmt.Errorf("%s: promotion target is empty", info.Filename))
+		return nil
+	}
+	target := v.targets.get(key)
+	for i := range config.Images {
+		if err := target.addStep(&config.Images[i], config.Images[i].Optional, excluded, included, info, key); err != nil {
+			v.errs = append(v.errs, fmt.Errorf("%s: %v", info.Filename, err))
+			return nil
+		}
+	}
+	for _, step := range config.RawSteps {
```

Comment:
This is not valid. You *must* list an image built in a raw step in the additional images for it to be published.